src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -89,6 +91,83 @@ private void WriteObjectOrArrayStart(ClassType classType, JsonEncodedText proper
             }
         }
 
+        public void WritePreservedObjectOrArrayStart(ClassType classType, Utf8JsonWriter writer, JsonSerializerOptions options, string referenceId)
+        {
+            if (JsonPropertyInfo?.EscapedName.HasValue == true)
+            {
+                WritePreservedObjectOrArrayStart(classType, JsonPropertyInfo.EscapedName!.Value, writer, referenceId);
+            }
+            else if (KeyName != null)
+            {
+                JsonEncodedText propertyName = JsonEncodedText.Encode(KeyName, options.Encoder);
+                WritePreservedObjectOrArrayStart(classType, propertyName, writer, referenceId);
+            }
+            else
+            {
+                if ((classType & (ClassType.Object | ClassType.Dictionary)) != 0)
+                {
+                    writer.WriteStartObject();
+                    writer.WriteString("$id", referenceId);
```

Comment:
We should store constant string literals like `$id` in static `JsonEncodedText` and write those (it's faster).

Reply:
Where should those go? My first guess is `WriteStack`